google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/SpannerTableExporter.java

```diff
@@ -40,94 +34,51 @@
   }
 
-  private String[] getTableString(Table table, Metadata metadata, Iterable<Column> keyColumns) {
-
-    String primaryKeyColNames = StreamSupport.stream(keyColumns.spliterator(), false)
-        .map(Column::getQuotedName)
-        .collect(Collectors.joining(","));
-
-    StringJoiner colsAndTypes = new StringJoiner(",");
-
-    ((Iterator<Column>) table.getColumnIterator()).forEachRemaining(col -> colsAndTypes
-        .add(col.getQuotedName()
-            + " " + col.getSqlType(this.spannerDialect, metadata)
-            + (col.isNullable() ? this.spannerDialect.getNullColumnString() : " not null")));
-
-    ArrayList<String> statements = new ArrayList<>();
-    statements.add(MessageFormat.format(this.createTableTemplate, table.getQuotedName(),
-        colsAndTypes.toString(), primaryKeyColNames));
-
-    // Hibernate requires the special hibernate_sequence table to be populated with an initial val.
-    if (table.getName().equals(SequenceStyleGenerator.DEF_SEQUENCE_NAME)) {
-      statements.add("INSERT INTO " + SequenceStyleGenerator.DEF_SEQUENCE_NAME + " ("
-          + SequenceStyleGenerator.DEF_VALUE_COLUMN + ") VALUES(1)");
-    }
-
-    return statements.toArray(new String[0]);
+  @Override
+  public String[] getSqlDropStrings(Table currentTable, Metadata metadata) {
+    return buildSqlStrings(currentTable, metadata, false);
   }
 
-  private Table getContainingTableForCollection(Metadata metadata, Table collectionTable) {
-    for (Collection collection : metadata.getCollectionBindings()) {
-      if (collection.getCollectionTable().equals(collectionTable)) {
-        return collection.getTable();
-      }
-    }
-    return null;
+  /**
+   * Initializes the table exporter for if a new create-table or drop-table sequence is starting.
+   */
+  public void initializeTableExporter(Metadata metadata, boolean isCreateTables) {
+    tableDependencyTracker.initializeDependencies(metadata, isCreateTables);
   }
 
-  @Override
-  public String[] getSqlDropStrings(Table table, Metadata metadata) {
-    /* Cloud Spanner requires examining the metadata to find all indexes and interleaved tables.
-     * These must be dropped before the given table can be dropped.
-     * The current implementation does not support interleaved tables.
-     */
-
-    ArrayList<String> dropStrings = new ArrayList<>();
+  private String[] buildSqlStrings(Table currentTable, Metadata metadata, boolean isCreateTables) {
+    Collection<Table> tablesToProcess = tableDependencyTracker.getDependentTables(currentTable);
 
-    Iterator<Index> iteratorIdx = table.getIndexIterator();
-    while (iteratorIdx.hasNext()) {
-      Index curr = iteratorIdx.next();
-      dropStrings.add("drop index " + curr.getName());
-    }
+    List<String> ddlStatements = tablesToProcess.stream()
+        .flatMap(table -> {
+          if (isCreateTables) {
+            return spannerTableStatements.createTable(table, metadata).stream();
```

Comment:
since both `createTable()` and `dropTable()` just return a list, you may be able to:
* replace `.flatMap()` with `.map()`
* remove `.stream()` 
* remove `.collect()`

Reply:
Discussed offline; won't work because the operation needs to flatten nested lists.

Reply:
Will stick with the flatmap method as discussed offline.